### ISSUE TITLE
Workspaces not restoring properly

### ISSUE THREAD

**Issue by mellery451** [vim-ctrlspace/vim-ctrlspace#202]
I'm using 5 or 6 tabs to organize a few dozen files. This approach has worked flawlessly for a few years.  Lately I've noticed that the workspace often does not restore properly...I'll typically have one or two tabs that have lost their name (that I had assigned the tab group) and/or buffers will not be open in the relevant group. I have lots of other plugins installed, so it's certainly possible there is a bad interaction going on, but it also feels to me like this started sometime around my switch to vim-8 (homebrew version). 

Anyone else having issues saving/restoring custom tab groups properly?

**Comment by cometsong** [comment]
I've also just barely noticed a tab losing the name I gave it when restored.  And also just updated my installed vim version on my mac as well as on a centos server which I work.  I'm going to check with older vim versions using my otherwise-identical config to see if I can find the problem source.  Might go back a few steps of recent option/plugin change on my end too, to test those.

**Comment by mellery451** [comment]
In very limited testing on my part, it seems like the issue does not manifest with neovim (v0.3.1 from homebrew) whereas pretty consistently happens with vim 8.1 - using the same init/vimrc in both cases. Just another data point for consideration.

**Comment by filedil** [comment]
I believe at some point Vim changed the way it writes sessions (with mksession). Possibly in relation to changelog item "8.1.0149  session is wrong with multiple tabs when :lcd was used".
I've looked at the "ctrlspace#workspaces#SaveWorkspace" function and the the loop over CS_SESSION file is looking for wrong commands.

I have compared session files between Vim 8.1.0229 and 8.0.1542.
Two tabs, 1 buffer visible in each tab, 8 buffers total.
The older Vim session goes roughly like this:
```
...
badd file1.ext
badd fileX.ext (x7)
...
edit file4.ext
...
tabedit file8.ext
...
```

The new one is different:
```
...
badd file1.ext
badd fileX.ext (x7)
...
tabnew
tabnext -1
edit file4.ext
...
tabnext
edit file8.ext
...
tabnext 1
...
```
So with 'edit' after 'tabnew' and switching using 'tabnext', it breaks the order of operations that Ctrlspace expects.

**Comment by filedil** [comment]
For now, as a workaround, I changed this line in autoload/ctrlspace/workspaces.vim:
`elseif ((cmd =~# "^edit") && (tabIndex == 0)) || (cmd =~# "^tabnew") || (cmd =~# "^tabedit")`
to:
`elseif ((cmd =~# "^edit") && (tabIndex == 0)) || (cmd =~# "^tabnext$")`.
And everything seems to be working correctly. 
The correct way would probably be to detect the "version" of session file and react accordingly.

**Comment by cometsong** [comment]
Thanks @filedil !

I'm starting to research the session-saving situation within vim versions.
Looking through commits as well as checking session files in different installations.
It will not be a fast task, but it's in progress.  🙂

**Comment by Konfekt** [comment]
Hello, I can confirm this issue appearing since switching from Vim 8.0.1568 to 8.1.0892.

**Comment by Konfekt** [comment]
In the meanwhile, wouldn't simply checking whether the  version of the instance of Vim running  (instead of that which created the session file)  is `< 8.1.0149`be a good workaround?

**Comment by Konfekt** [comment]
@SunsetCat  The fix that you propose in https://github.com/vim-ctrlspace/vim-ctrlspace/issues/202#issuecomment-410082543 changes the way `CtrlSpace` saves the workspace, isnt'it? Therefore, the workspace has to be written with the patched code, and then the workspace opens fine? It does not work over here.

**Comment by Konfekt** [comment]
For reference, I replaced
```
elseif ((cmd =~# "^edit") && (tabIndex == 0)) || (cmd =~# "^tabnew$") || (cmd =~# "^tabedit")
```
by
```vim
elseif ((cmd =~# "^edit") && (tabIndex == 0))
					\ || cmd =~# (has('patch-8.1.149') ?
					\            "^tabnnext$" : "^tabnew$\|^tabedit")
```

**Comment by Konfekt** [comment]
Dear @szw, do you have any idea on how to tackle this?

**Comment by Konfekt** [comment]
@cometsong Did the solution found by @SunsetCat solve this issue for you?

**Comment by cometsong** [comment]
@Konfekt - Sorry not earlier reply - I've been testing combinations of the `tab`-specific tests along with the patch test, and some of them work on the later version, some on the earlier, but none yet on both.  Now I'm looking at option to re-make the whole `if-else` clause for this many checks in a row. 🤔

**Comment by cometsong** [comment]
```There can be only one``` *patch-level* ```!```  - Connor MacLeod  (mostly.)

I've got it working using a two-part check; either it `has` or `!has` the patch.

It works in four diff versions of `vim` on my end... 
Can you (please) check it on yours using one version pre-`8.1.149` and another post-patch?

The `elseif` clause is here:

```vim
elseif
    \    ((cmd =~# "^edit") && (tabIndex == 0))
    \ || (!has('patch-8.1.149') && ( (cmd =~# "^tabnew") || (cmd =~# "^tabedit") ))
    \ ||  (has('patch-8.1.149') && (cmd =~# "^tabnext$"))
```

**Comment by Konfekt** [comment]
Works post 8.1.149.

**Comment by cometsong** [comment]
Excellent.  I'll be committing and releasing this fix later today.

**Comment by cometsong** [comment]
Actually, I haven't tested it with `neovim`, so it's on hold until I actually install that again for testing.
(Not using it due to it's absence on all my remote servers.)

**Comment by cometsong** [comment]
Solved in 1a285ba. 
The saving/loading of Workspaces now works pre/post 8.1.149, plus with nvim v0.3.4.
